src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -8,582 +8,198 @@
-                    else:
-                        raise NotImplementedError()
-                consumed_domains.append(consumed_domain)
-
-            # compute the bounds of all consumed domains
-            if consumed_domains:
-                if all(
-                    consumed_domain.ranges.keys() == consumed_domains[0].ranges.keys()
-                    for consumed_domain in consumed_domains
-                ):  # scalar otherwise
-                    domains[param] = domain_union(*consumed_domains).as_expr()
-
-    return FencilWithTemporaries(
-        fencil=ir.FencilDefinition(
-            id=node.fencil.id,
-            function_definitions=node.fencil.function_definitions,
-            params=node.fencil.params[:-1],  # remove `_gtmp_auto_domain` param again
-            closures=list(reversed(closures)),
-            location=node.fencil.location,
-            implicit_domain=node.fencil.implicit_domain,
+        # extract if_ call to the top-level
+        functools.partial(
+            _transform_by_pattern, predicate=lambda expr, _: cpm.is_call_to(expr, "if_")
         ),
-        params=node.params,
-        tmps=node.tmps,
-    )
-
+    ]
 
-def _tuple_constituents(node: ir.Expr) -> Iterable[ir.Expr]:
-    if cpm.is_call_to(node, "make_tuple"):
-        for arg in node.args:
-            yield from _tuple_constituents(arg)
-    else:
-        yield node
-
-
-def collect_tmps_info(
-    node: FencilWithTemporaries, *, offset_provider: common.OffsetProvider
-) -> FencilWithTemporaries:
-    """Perform type inference for finding the types of temporaries and sets the temporary size."""
-    tmps = {tmp.id for tmp in node.tmps}
-    domains: dict[str, ir.Expr] = {}
-    for closure in node.fencil.closures:
-        for output_field in _tuple_constituents(closure.output):
-            assert isinstance(output_field, ir.SymRef)
-            if output_field.id not in tmps:
-                continue
-
-            assert output_field.id not in domains or domains[output_field.id] == closure.domain
-            domains[output_field.id] = closure.domain
-
-    new_node = FencilWithTemporaries(
-        fencil=node.fencil,
-        params=node.params,
-        tmps=[
-            ir.Temporary(id=tmp.id, domain=domains[tmp.id], dtype=tmp.dtype) for tmp in node.tmps
-        ],
-    )
-    # TODO(tehrengruber): type inference is only really needed to infer the types of the temporaries
-    #  and write them to the params of the inner fencil. This should be cleaned up after we
-    #  refactored the IR.
-    return itir_type_inference.infer(new_node, offset_provider=offset_provider)
+    while unprocessed_stmts:
+        stmt = unprocessed_stmts.pop(0)
 
+        did_transform = False
+        for transform in transforms:
+            transformed_stmts = transform(stmt=stmt, declarations=declarations, uids=uids)
+            if transformed_stmts:
+                unprocessed_stmts = [*transformed_stmts, *unprocessed_stmts]
```

Comment:
Using a [`collections.deque`](https://devdocs.io/python~3.13/library/collections#collections.deque) for `unprocessed_stmts` could make this simpler and probably more efficient.
```suggestion
                unprocessed_stmts.appendleft(transformed_stmts)
```

Reply:
A deque doesn't really fit since
1. only need a one sided queue
2. `deque.extendleft` reverses the order `transformed_stmts` which is wrong here. Manually reversing nulls simplicity again.
I can reverse the order of the list if you prefer that, but it feels like premature optimizations.